src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs

```diff
@@ -742,35 +743,163 @@ public override ClaimsPrincipal ValidateToken(string token, TokenValidationParam
+                            claimsPrincipal = outerToken != null ? ValidateJWE(token, outerToken, validationParameters, currentConfiguration, out signatureValidatedToken, out exceptionThrown) :
+                                ValidateJWS(token, validationParameters, currentConfiguration, out signatureValidatedToken, out exceptionThrown);
+
+                            if (claimsPrincipal != null)
+                                return claimsPrincipal;
+                        }
+                    }
+
+                    // If we were still unable to validate, attempt to refresh the configuration and validate using it
+                    // but ONLY if the currentConfiguration is not null. We want to avoid refreshing the configuration on
+                    // retrieval error as this case should have already been hit before. This refresh handles the case
+                    // where a new valid configuration was somehow published during validation time.
+                    if (currentConfiguration != null)
+                    {
+                        validationParameters.ConfigurationManager.RequestRefresh();
+                        var lastConfig = currentConfiguration;
+                        currentConfiguration = validationParameters.ConfigurationManager.GetBaseConfigurationAsync(CancellationToken.None).GetAwaiter().GetResult();
+
+                        // Only try to re-validate using the newly obtained config if it doesn't reference equal the previously used configuration.
+                        if (lastConfig != currentConfiguration)
+                            claimsPrincipal = outerToken != null ? ValidateJWE(token, outerToken, validationParameters, currentConfiguration, out signatureValidatedToken, out exceptionThrown) :
+                                ValidateJWS(token, validationParameters, currentConfiguration, out signatureValidatedToken, out exceptionThrown);
+                    }
+                }
+            }
+
+            if (claimsPrincipal != null)
+                return claimsPrincipal;
+
+            throw exceptionThrown;
+        }
+
+        private ClaimsPrincipal ValidateJWE(string decryptedJwt, JwtSecurityToken outerToken, TokenValidationParameters validationParameters, BaseConfiguration currentConfiguration, out SecurityToken signatureValidatedToken, out Exception exceptionThrown)
+        {
+            exceptionThrown = null;
+            signatureValidatedToken = null;
+            try
+            {
+                var innerToken = ValidateSignature(decryptedJwt, validationParameters, currentConfiguration);
+                outerToken.InnerToken = innerToken;
+                signatureValidatedToken = outerToken;
+                return ValidateTokenPayload(innerToken, validationParameters, currentConfiguration);
+            }
+            catch (Exception ex)
+            {
+                exceptionThrown = ex;
+                return null;
+            }
```

Comment:
I'm not crazy about this sort of pattern in general (catching generic exceptions and then assuming the caller will always throw in the right way) but if we go down this route at least consider using ExceptionDispatchInfo instead. I believe the way the code is currently written, it will strip the stack trace info which I don't think is the intent and hurts debugging efforts.

[https://docs.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo.capture?redirectedfrom=MSDN&view=net-6.0#System_Runtime_ExceptionServices_ExceptionDispatchInfo_Capture_System_Exception](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo.capture?redirectedfrom=MSDN&view=net-6.0#System_Runtime_ExceptionServices_ExceptionDispatchInfo_Capture_System_Exception)_

Reply:
I like this a lot better than what was there before but it really begs the question for me, why are we even passing anything around for the exception, if we're just throwing in the method just above in the call stack why not just do a regular throw here instead of capturing it, letting control return to the caller and then immediately throwing?

Reply:
Ah I see, it was too long since I looked at this last. It's to allow for the fallback in logic. This makes sense, this part LGTM.